File: Mobile/LifeOrganizerMobile/src/types/habit.ts
import { DayOfWeek } from "./days";

export enum HabitFrequency {
    Daily = 0,
    Weekly = 1,
    Monthly = 2,
    Custom = 3,
}

export enum HabitCompletionStatus {
    Completed = 0,
    Missed = 1,
}

export type Habit = {
    id: string;
    name: string;
    frequency: HabitFrequency;
    scheduledDays: DayOfWeek[];
    completionDeadline?: string; // TimeSpan - "HH:mm:ss"
    isActive: boolean;
    createdAt: string;
    isCompletedToday?: boolean;
}

export type HabitCompletion = {
    date: string; // DateOnly - "yyyy-MM-dd"
    status: HabitCompletionStatus;
};

export type HabitDetails = {
    id: string;
    name: string;
    frequency: HabitFrequency;
    scheduledDays: DayOfWeek[];
    completionDeadline?: string;
    recentCompletions: HabitCompletion[];
};

File: Server/LifeOrganizer/LifeOrganizer.Domain/Entities/Habit.cs
using LifeOrganizer.Domain.Enums;

namespace LifeOrganizer.Domain.Entities
{
    public class Habit
    {
        public Guid Id { get; set; }
        public Guid UserId { get; set; }
        public User User { get; set; } = null!;
        public string Name { get; set; } = string.Empty;
        public HabitFrequency Frequency { get; set; }
        public List<DayOfWeek> ScheduledDays { get; set; } = new();
        public TimeSpan? CompletionDeadline { get; set; }
        public bool IsActive { get; set; } = true;
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public ICollection<HabitCompletion> Completions { get; set; } = new List<HabitCompletion>();
        public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
    }
}
